docs/configuration-guide.md
# Configuration Guide

The confd configuration file is written in [TOML](https://github.com/mojombo/toml)
and loaded from `/etc/confd/confd.toml` by default. You can specify the config file via the `-config-file` command line flag.

> Note: You can use confd without a configuration file. See [Command Line Flags](https://github.com/kelseyhightower/confd/blob/master/docs/command-line-flags.md).

Optional:

* `backend` (string) - The backend to use. ("etcd")
* `client_cakeys` (string) - The client CA key file.
* `client_cert` (string) - The client cert file.
* `client_key` (string) - The client key file.
* `confdir` (string) - The path to confd configs. ("/etc/confd")
* `interval` (int) - The backend polling interval in seconds. (600)
* `log-level` (string) - level which confd should log messages ("info")
* `nodes` (array of strings) - List of backend nodes. (["http://127.0.0.1:4001"])
* `noop` (bool) - Enable noop mode. Process all template resources; skip target update.
* `prefix` (string) - The string to prefix to keys. ("/")
* `scheme` (string) - The backend URI scheme. ("http" or "https")
* `srv_domain` (string) - The name of the resource record.
* `srv_record` (string) - The SRV record to search for backends nodes.
* `sync-only` (bool) - sync without check_cmd and reload_cmd.
* `watch` (bool) - Enable watch support.
* `auth_token` (string) - Auth bearer token to use.
* `auth_type` (string) - Vault auth backend type to use.
* `basic_auth` (bool) - Use Basic Auth to authenticate (only used with -backend=consul and -backend=etcd).
* `table` (string) - The name of the DynamoDB table (only used with -backend=dynamodb).
* `separator` (string) - The separator to replace '/' with when looking up keys in the backend, prefixed '/' will also be removed (only used with -backend=redis)
* `username` (string) - The username to authenticate as (only used with vault and etcd backends).
* `password` (string) - The password to authenticate with (only used with vault and etcd backends).
* `app_id` (string) - Vault app-id to use with the app-id backend (only used with -backend=vault and auth-type=app-id).
* `user_id` (string) - Vault user-id to use with the app-id backend (only used with -backend=value and auth-type=app-id).
* `role_id` (string) - Vault role-id to use with the AppRole, Kubernetes backends (only used with -backend=vault and either auth-type=app-role or auth-type=kubernetes).
* `secret_id` (string) - Vault secret-id to use with the AppRole backend (only used with -backend=vault and auth-type=app-role).
* `file` (array of strings) - The YAML file to watch for changes (only used with -backend=file).
* `filter` (string) - Files filter (only used with -backend=file) (default "*").
* `path` (string) - Vault mount path of the auth method (only used with -backend=vault).

Example:

```TOML
backend = "etcd"
client_cert = "/etc/confd/ssl/client.crt"
client_key = "/etc/confd/ssl/client.key"
confdir = "/etc/confd"
log-level = "debug"
interval = 600
nodes = [
  "http://127.0.0.1:4001",
]
noop = false
prefix = "/production"
scheme = "https"
srv_domain = "etcd.example.com"
```


CHANGELOG
### v0.16.0

28f72608 Release arm64 binaries
4528af01 Add option to specify auth backend mount path in Vault
b51998d6 new etcdv3 implementation
fa37c993 isFileExist() was not defined in windows
a23aa1c7 Update Vault Kubernetes authentication doc
9b0ce8e6 Fix vendor name for logrus
ccc17eb5 fix fileStat close when open failed on windows
e3864219 add parseBool to template_funcs
94bf80e7 Add Kubernetes auth for Vault backend
304efb5a Add AppRole authN for Vault backend
fb208461 Redis Backend Updates: WatchPrefix Support, Hash Support, Standard Redis Key Naming Convention
36d874f5 Fix bug with multiple watches for etcdv3
1fd12493 backend/etcd: add scheme to srv discovered endpoints
a88c922a Vault TLS certificates auth method
b77b9232 Add atoi function to template funcs and test

### v0.15.0

0c4cb29e Don't pass empty prefix to zookeeper nodeWalk
5d6e2790 Update template.md - added forgotten verb
19d6514d remove fmt.Println
802e6a11 Added windows cmd support
59001ca3 Fix handling of absent parameters for ssm backend
8419c708 AWS SSM - don't treat "ParameterNotFound" as a fatal error
0a09632a Add support for consul BasicAuth
adf8d486 Fix panic during Consul client creation
027be91f Update docker Multi-Stage build
376d3f99 fix etcdv3 client watch: reuse previous watch to avoid memory leak
f711530d fix(confd): etcd curl err when make integration

### v0.14.0

95ba861f Replace godep with dep
cb53644c Template integration with https://github.com/xordataexchange/crypt provided values
6da43945 add go version and git commit sha1 to version string
8f052819 yaml/json file backend
61652a2c Add AWS SSM Parameter Store backend

### v0.13.0

181e3e3 Update dependencies
509268d Remove StackEngine backend
3a61969 Add further installation instructions
dd47bbb Introducing WaitPrefix for the Rancher backend
35ffe8f Adding nested key support to Vault backend v2
f1833a6 ETCD V3 support
89c59d4 Feature new template func add
2b080af added some additional text/template sorting methods
fbea3b9 Fixes #562 by not adding the URL scheme to the node at DNS lookup time
ab93ff1 Windows support
695e62e add trimSuffix template function
3049df8 Add base64 encode/decode template helpers
485b31c improved etcd watch support
8528d3c added support to specify a redis database (in the connection string)

### v0.12.0

5f3f79d Change etcd node port because the official port of etcd is 2379
9073dae Fix typo in docs
0ea1edb Remove unnecessary loop in Consul backend.
dbbee0a Fix travis configuration to use latest dynamoDB binary
ad5f74f Fix typo in quick-start-guide.md
7ac1091 Upgrade Dockerfile to use latest stable Alpine which has Go 1.6
20b3d37 add env backend integration tests
c6622ed updated template docs
9d57489 Specify default Vault backend node
b34d223 Add simple redis example since one was missing from quickstart
32d46e5 Fix typo in template-resources.md
7c89004 Fix up Rancher metadata service description and links
2a54c58 Fix missing braces and extra parenthesis on formatted datatime example
6e198db Fix DynamoDB credentials handling
705f2ee Fix dynamodb nil pointers
e48e48f Use SVG for the build image
6bb3c21 Add file exists function to template functions
57b649d zookeeper watch support
59da585 extended getenv & map template functions
d423123 updated README for vault
e60c2e2 Add support for redis password authentication
f3a02e7 Adding example goreman Procfile for etcd
aae4190 updated Dockerfile.build.alpine
cf62311 update some docs
54f715c Add lookupSRV function.
7afb8d6 fixed broken etcd watch support
5a5af59 switched to new etcd client
5d2955a Remove gb install
1577855 Fix typos in docs/tomcat-sample.md
d133c25 Added support for Vault backend
071e6ab IP dns resolving function for templates
ccf2486 redis: try to recover from closed connections
66a31bf Add support for etcd BasicAuth
d9ea46c Added documentation for stackengine as a backend
a1dee21 add -sync-only flag
c615c18 update checklist
233ac99 Add more debugging output for onetime and env.
d7f835d Create tomcat-sample.md

### v0.11.0

46d3c69 load template resources every interval
6b01b27 Add Rancher Metadata backend
bec0b60 link to the right command-line-flags doc
be20ca7 Fix typo in templates.md
745cf2b add default node for zookeeper backend
0dd4c53 Add StackEngine backend support
f446ce4 Load AWS credentials through a provider chain
f206b47 Add strings.Replace support

### v0.10.0

70cf76a move contrib to root dir
34a0ff7 Updating README.md to include new backends
fee8361 Fix broken links to documentation
d9d4cb1 Use the gb tool for dep management and building.
1ef34cb template: don't crash on template error
7531530 Log command output on error
fed32fe Draft: Support for DynamoDB
7f59d50 Use correct application name for ZooKeeper example
03a7c7e feat: add strings.Contains support
6307c26 change opdemand emails to engine yard
5350113 client cert config from environment variables
b6f414e Adding better support for Consul, enabling SSL/TLS client config
e94151b update release checklist
ce4eb5c bump to v1.0.0-dev
503d745 remove --quiet, --debug and --verbose
c9cefaf random doc fixes

### v0.9.0

8888b7a Add maintainers file
431e31b add release checklist document
f782957 bump etcd to v2.0.9 for integration tests
1b2d669 *feature) add recursive get toml files.
94fc0a6 allow override of BIN_DIR
7cce514 log that we're skipping config file to debug
6e0cbfc Enable watch from the config file
0410cbd bump version to 0.8.0
c7f53e1 add go 1.4 to travis.ci version matrix
7a2e2f4 Add toUpper, toLower support
4a8e079 Fix "The event in requested index is outdated and cleared" error
211a5cb add --log-level flag
0a0858c fix godeps
821cd61 properly format error messages
2ad79d6 general integration test nitpicks
913eb2e hook up etcd integration tests
2ad2fd4 Improved docs for how to use go templates to traverse json

config.go
package main

import (
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/BurntSushi/toml"
	"github.com/kelseyhightower/confd/backends"
	"github.com/kelseyhightower/confd/log"
	"github.com/kelseyhightower/confd/resource/template"
)

type TemplateConfig = template.Config
type BackendsConfig = backends.Config

// A Config structure is used to configure confd.
type Config struct {
	TemplateConfig
	BackendsConfig
	Interval      int    `toml:"interval"`
	SecretKeyring string `toml:"secret_keyring"`
	SRVDomain     string `toml:"srv_domain"`
	SRVRecord     string `toml:"srv_record"`
	LogLevel      string `toml:"log-level"`
	Watch         bool   `toml:"watch"`
	PrintVersion  bool
	ConfigFile    string
	OneTime       bool

}

var config Config

func init() {
	flag.StringVar(&config.AuthToken, "auth-token", "", "Auth bearer token to use")
	flag.StringVar(&config.Backend, "backend", "kubernetes", "backend to use, Default kubernetes")
	flag.BoolVar(&config.BasicAuth, "basic-auth", false, "Use Basic Auth to authenticate (only used with -backend=consul and -backend=etcd)")
	flag.StringVar(&config.ClientCaKeys, "client-ca-keys", "", "client ca keys")
	flag.StringVar(&config.ClientCert, "client-cert", "", "the client cert")
	flag.StringVar(&config.ClientKey, "client-key", "", "the client key")
        flag.BoolVar(&config.ClientInsecure, "client-insecure", false, "Allow connections to SSL sites without certs (only used with -backend=etcd)")
	flag.StringVar(&config.ConfDir, "confdir", "/etc/confd", "confd conf directory")
	flag.StringVar(&config.ConfigFile, "config-file", "/etc/confd/confd.toml", "the confd config file")
	flag.Var(&config.YAMLFile, "file", "the YAML file to watch for changes (only used with -backend=file)")
	flag.StringVar(&config.Filter, "filter", "*", "files filter (only used with -backend=file)")
	flag.IntVar(&config.Interval, "interval", 600, "backend polling interval")
	flag.BoolVar(&config.KeepStageFile, "keep-stage-file", false, "keep staged files")
	flag.StringVar(&config.LogLevel, "log-level", "", "level which confd should log messages")
	flag.Var(&config.BackendNodes, "node", "list of backend nodes")
	flag.BoolVar(&config.Noop, "noop", false, "only show pending changes")
	flag.BoolVar(&config.OneTime, "onetime", false, "run once and exit")
	flag.StringVar(&config.Prefix, "prefix", "", "key path prefix")
	flag.BoolVar(&config.PrintVersion, "version", false, "print version and exit")
	flag.StringVar(&config.Scheme, "scheme", "http", "the backend URI scheme for nodes retrieved from DNS SRV records (http or https)")
	flag.StringVar(&config.SecretKeyring, "secret-keyring", "", "path to armored PGP secret keyring (for use with crypt functions)")
	flag.StringVar(&config.SRVDomain, "srv-domain", "", "the name of the resource record")
	flag.StringVar(&config.SRVRecord, "srv-record", "", "the SRV record to search for backends nodes. Example: _etcd-client._tcp.example.com")
	flag.BoolVar(&config.SyncOnly, "sync-only", false, "sync without check_cmd and reload_cmd")
	flag.StringVar(&config.AuthType, "auth-type", "", "Vault auth backend type to use (only used with -backend=vault)")
	flag.StringVar(&config.AppID, "app-id", "", "Vault app-id to use with the app-id backend (only used with -backend=vault and auth-type=app-id)")
	flag.StringVar(&config.UserID, "user-id", "", "Vault user-id to use with the app-id backend (only used with -backend=value and auth-type=app-id)")
	flag.StringVar(&config.RoleID, "role-id", "", "Vault role-id to use with the AppRole, Kubernetes backends (only used with -backend=vault and either auth-type=app-role or auth-type=kubernetes)")
	flag.StringVar(&config.SecretID, "secret-id", "", "Vault secret-id to use with the AppRole backend (only used with -backend=vault and auth-type=app-role)")
	flag.StringVar(&config.Path, "path", "", "Vault mount path of the auth method (only used with -backend=vault)")
	flag.StringVar(&config.Table, "table", "", "the name of the DynamoDB table (only used with -backend=dynamodb)")
	flag.StringVar(&config.Separator, "separator", "", "the separator to replace '/' with when looking up keys in the backend, prefixed '/' will also be removed (only used with -backend=redis)")
	flag.StringVar(&config.Username, "username", "", "the username to authenticate as (only used with vault and etcd backends)")
	flag.StringVar(&config.Password, "password", "", "the password to authenticate with (only used with vault and etcd backends)")
	flag.BoolVar(&config.Watch, "watch", false, "enable watch support")
}

// initConfig initializes the confd configuration by first setting defaults,
// then overriding settings from the confd config file, then overriding
// settings from environment variables, and finally overriding
// settings from flags set on the command line.
// It returns an error if any.
func initConfig() error {
	_, err := os.Stat(config.ConfigFile)
	if os.IsNotExist(err) {
		log.Debug("Skipping confd config file.")
	} else {
		log.Debug("Loading " + config.ConfigFile)
		configBytes, err := ioutil.ReadFile(config.ConfigFile)
		if err != nil {
			return err
		}

		_, err = toml.Decode(string(configBytes), &config)
		if err != nil {
			return err
		}
	}

	// Update config from environment variables.
	processEnv()

	if config.SecretKeyring != "" {
		kr, err := os.Open(config.SecretKeyring)
		if err != nil {
			log.Fatal(err.Error())
		}
		defer kr.Close()
		config.PGPPrivateKey, err = ioutil.ReadAll(kr)
		if err != nil {
			log.Fatal(err.Error())
		}
	}

	if config.LogLevel != "" {
		log.SetLevel(config.LogLevel)
	}

	if config.SRVDomain != "" && config.SRVRecord == "" {
		config.SRVRecord = fmt.Sprintf("_%s._tcp.%s.", config.Backend, config.SRVDomain)
	}

	// Update BackendNodes from SRV records.
	if config.Backend != "env" && config.SRVRecord != "" {
		log.Info("SRV record set to " + config.SRVRecord)
		srvNodes, err := getBackendNodesFromSRV(config.SRVRecord)
		if err != nil {
			return errors.New("Cannot get nodes from SRV records " + err.Error())
		}

		switch config.Backend {
		case "etcd":
			vsm := make([]string, len(srvNodes))
			for i, v := range srvNodes {
				vsm[i] = config.Scheme + "://" + v
			}
			srvNodes = vsm
		}

		config.BackendNodes = srvNodes
	}
	if len(config.BackendNodes) == 0 {
		switch config.Backend {
		case "consul":
			config.BackendNodes = []string{"127.0.0.1:8500"}
		case "etcd":
			peerstr := os.Getenv("ETCDCTL_PEERS")
			if len(peerstr) > 0 {
				config.BackendNodes = strings.Split(peerstr, ",")
			} else {
				config.BackendNodes = []string{"http://127.0.0.1:4001"}
			}
		case "etcdv3":
			config.BackendNodes = []string{"127.0.0.1:2379"}
		case "redis":
			config.BackendNodes = []string{"127.0.0.1:6379"}
		case "vault":
			config.BackendNodes = []string{"http://127.0.0.1:8200"}
		case "zookeeper":
			config.BackendNodes = []string{"127.0.0.1:2181"}
		case "kubernetes":
			if len(config.Kubeconfig) == 0{
				config.BackendNodes = []string{"https://kubernetes.default"}
				config.InCluster = true
			}

		}
	}
	// Initialize the storage client
	log.Info("Backend set to " + config.Backend)

	if config.Watch {
		unsupportedBackends := map[string]bool{
			"dynamodb": true,
			"ssm":      true,
		}

		if unsupportedBackends[config.Backend] {
			log.Info(fmt.Sprintf("Watch is not supported for backend %s. Exiting...", config.Backend))
			os.Exit(1)
		}
	}

	if config.Backend == "dynamodb" && config.Table == "" {
		return errors.New("No DynamoDB table configured")
	}
	config.ConfigDir = filepath.Join(config.ConfDir, "conf.d")
	config.TemplateDir = filepath.Join(config.ConfDir, "templates")
	return nil
}

func getBackendNodesFromSRV(record string) ([]string, error) {
	nodes := make([]string, 0)

	// Ignore the CNAME as we don't need it.
	_, addrs, err := net.LookupSRV("", "", record)
	if err != nil {
		return nodes, err
	}
	for _, srv := range addrs {
		host := strings.TrimRight(srv.Target, ".")
		port := strconv.FormatUint(uint64(srv.Port), 10)
		nodes = append(nodes, net.JoinHostPort(host, port))
	}
	return nodes, nil
}

func processEnv() {
	cakeys := os.Getenv("CONFD_CLIENT_CAKEYS")
	if len(cakeys) > 0 && config.ClientCaKeys == "" {
		config.ClientCaKeys = cakeys
	}

	cert := os.Getenv("CONFD_CLIENT_CERT")
	if len(cert) > 0 && config.ClientCert == "" {
		config.ClientCert = cert
	}

	key := os.Getenv("CONFD_CLIENT_KEY")
	if len(key) > 0 && config.ClientKey == "" {
		config.ClientKey = key
	}
}
